### ISSUE TITLE
Notify developers again about pending rejection as they get close to the deadline

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#14814]
From discussion in https://github.com/mozilla/addons-server/issues/14454#issuecomment-649548842 we want to notify developers about pending rejections as they get close to the deadline, if they haven't submitted a new version to fix the issues. 

This will eventually replace the cron that notifies developers about expiring info requests, but we can probably remove that in a separate issue later. The two can coexist, because once we change reviewer tools to allow delayed rejections, we'll also stop adding deadlines to info requests.

**Comment by diox** [comment]
Note: the info requests were per add-on, the delayed rejections will be per-version, but this shouldn't be a problem, we can group by add-on as multiple version rejections coming from the same reviewer action should have the same deadline.

**Comment by diox** [comment]
A couple cases where we _won't_ send that notification:
- If the versions are already rejected or deleted
- If the add-on is already disabled or deleted
- If the developer(s) have already been notified since the last time they had a version scheduled for rejection
- If the add-on latest (non-disabled, non-deleted) version is not pending rejection

**Comment by AlexandraMoga** [comment]
@diox what is 'close to deadline' more exactly, 1 day before the grace period expires? 

Also, I'm not entirely sure what this means:
>A couple cases where we won't send that notification:
>* If the developer(s) have already been notified since the last time they had a version scheduled for rejection

**Comment by diox** [comment]
> what is 'close to deadline' more exactly, 1 day before the grace period expires?

It's 1 day yes.

> A couple cases where we won't send that notification: If the developer(s) have already been notified since the last time they had a version scheduled for rejection

This was poorly worded :) Consider the following scenario:
- On August 4, reviewer puts version A on delayed rejection for August 11
- On August 5, Reviewer puts version B on delayed rejection for August 12

In that case, we only send the last warning notification once on August 10, for Version A - it's an unlikely scenario, but we don't want to overload the reviewer with notifications (especially as they'll receive the email when the delayed rejection is set anyway). So we never notify about version B.

Now consider this other scenario:
- On August 4, reviewer puts version A on delayed rejection for August 11
- On August 10, we send the warning notification about version A
- Right after we send the notification, reviewer puts version B on delayed rejection for August 17
- On August 16, we _do_ send the notification about version B

This time because the second delayed rejection about version B happened *after* we sent the notification about version A, we will send the notification about version B later.

Note: for both scenarios the other rules continue to apply, so the state of the latest non-disabled non-deleted version also matters.

**Comment by AlexandraMoga** [comment]
These are the results for a set of add-ons for which a pending rejection of 2 days was added on August 4 and the last warning notification (if applicable) was expected to come on August 5:

The results were as I expected (for -dev): 
- last warning notification sent:
  - add-ons latest public version is pending rejection and there is no other listed version submitted in the meantime
  - add-on with mixed versions: listed version pending rejection while a new unlisted version is available
  - add-on's latest public version is pending rejection  while the most recent listed version is disabled
- last warning notification not sent:
  - add-on's latest public version is pending rejection but a new listed version is Awaiting review
  - add-on is forced disabled while having a pending rejection
  - add-on version is fully rejected while having a pending rejection
  - add-on version with pending rejection is disabled or deleted by developer
  - add-on with versions pending rejection is deleted entirely

I still need to wait another day to verify the results for the more particular cases mentioned in https://github.com/mozilla/addons-server/issues/14814#issuecomment-668610945

**Comment by AlexandraMoga** [comment]
I also have the results for the remaining scenarios (although I feel like these need some more testing, especially the second one):

> * On August 4, reviewer puts version A on delayed rejection for August 11
> * On August 5, Reviewer puts version B on delayed rejection for August 12

Result: the notification is sent only once, for version A. In this [example](https://reviewers.addons-dev.allizom.org/en-US/reviewers/review/email-delayed-reject-multiple),  versions 1.7 and version 1.8. both had a delay rejection but we sent the notification only for 1.7.

@diox When you mentioned the following scenario, I believe you assumed both versions were already approved:
> * On August 4, reviewer puts version A on delayed rejection for August 11
> * On August 10, we send the warning notification about version A
> * Right after we send the notification, reviewer puts version B on delayed rejection for August 17
> * On August 16, we do send the notification about version B

My test was slightly different and I didn't get the expected results:
For this [add-on example](https://reviewers.addons-dev.allizom.org/en-US/reviewers/review/email-pndrej-after-first-notif), version 1.0 had a pending rejection of 2 days added on August 4. On August 5, the last warning notification was received. Then, a new version (1.1) was submitted which was also put on pending rejection with deadline on August 7. At this point, version 1.1 is Awaiting Review (auto-approval was disabled automatically when the first version was put on delay rejection). I expected to receive a new deadline notification on August 6 for version 1,1, but we did not send it, probably because version 1.1 was still Awaiting review at this point? 
There is another issue here though. I've mentioned that on August 5 the last warning notification for version 1.0 was sent - [first email](https://addons-internal.dev.mozaws.net/en-US/admin/models/amo/fakeemail/94778/change/). Then, after 2 more hours, another identical email was sent again also for version 1.0 - [second email](https://addons-internal.dev.mozaws.net/en-US/admin/models/amo/fakeemail/94785/change/). This hasn't occurred with other add-ons.

**Comment by diox** [comment]
> I expected to receive a new deadline notification on August 6 for version 1,1, but we did not send it, probably because version 1.1 was still Awaiting review at this point?

That's correct.

> There is another issue here though. I've mentioned that on August 5 the last warning notification for version 1.0 was sent - first email. Then, after 2 more hours, another identical email was sent again also for version 1.0 - second email. This hasn't occurred with other add-ons.

Right. This is because the (delayed) rejection of Version 1.1 was submitted in the meantime, and that reset the flag that tells us we've notified the developers - we don't know which version we've notified you about, we just know that we did notify... It's not ideal in this scenario, please file a follow-up for this.
